### PR TITLE
[REEF-1838] Add API to Kill Yarn Job Application from .Net client

### PR DESCRIPTION
**Pull request by jwang98052**
* Add API in IYarnREEFClient
* Implement in YarnClient and YarnREEFClient
* The  code is from #1352 and implemented on top of the latest interface change

JIRA: [REEF-1838](https://issues.apache.org/jira/browse/REEF-1838)
This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs [hunk 1]
```diff
@@ -33,5 +34,13 @@ public interface IYarnREEFClient : IREEFClient
...
         [Unstable("0.17", "Working in progress for rest API id returned")]
         Task<IReadOnlyDictionary<string, IApplicationReport>> GetApplicationReports();
+
+        /// <summary>
+        /// Kills the job application and return Job status
+        /// </summary>
+        /// <param name="appId"></param>
+        /// <returns></returns>
+        [Unstable("0.17", "Working in progress for rest API status returned")]
+        Task<FinalState> KillJobApplication(string appId);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: How does this behave if `appId` is invalid or refers to a job that never existed or has already ended?

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/IYarnRMClient.cs [hunk 2]
```diff
@@ -62,5 +62,18 @@ internal interface IYarnRMClient
...
+        /// </summary>
+        /// <param name="appId">The application identifier.</param>
+        void KillApplicationAsync(string appId);
+
+        /// <summary>
+        /// Kills the application asynchronous.
+        /// </summary>
+        /// <param name="appId">The application identifier.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        void KillApplicationAsync(string appId, CancellationToken cancellationToken);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/IYarnRMClient.cs]: Same as above: What's the semantics of this call under common conditions?

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloTask.cs [hunk 3]
```diff
@@ -16,6 +16,7 @@
 // under the License.
 
 using System;
+using System.Threading;
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloTask.cs]: Is this change necessary?

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs [hunk 4]
```diff
@@ -115,6 +114,17 @@ private void LogApplicationReport()
         }
 
         /// <summary>
+        /// Kill Job Application.
+        /// </summary>
+        /// <param name="appId"></param>
+        private void KillJobApplication(string appId)
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: Why do we need that method in HelloREEF? Where is it called from?

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: If AppId is invalid, or if the application has completed,  exception will be thrown from the rest call. We can add code to catch the exception, log it, but I would still like to throw it to the caller for caller to get details of the failure. The reason of the failure is wrapped in the exception details thrown from Yarn.

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/IYarnRMClient.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/IYarnRMClient.cs]: Same

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloTask.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloTask.cs]: Forgot to remove after testing.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: I called it after submitting the job in testing and removed call. I can leave the all and comment it out to show an example for the client.

Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: It is a bit confusing to have that method in a user example - and or C# version of HelloREEF is already quite long for a hello world application IMO. I would remove it from HelloREEF and write a separate unit test to check that kill actually works.

**Comment by motus** [comment]
Also, what do we do with PR #1352? It apparently implements the same functionality.

**Comment by jwang98052** [comment]
@motus, This PL is built on IYarnREEFClient interface we introduced, with bug fixes and testing added that will replace  #1352 which should be closed.

**Comment by jwang98052** [comment]
I have pushed the update. Please have another look.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: We don't run .Net test on Yarn. To avoid polluting simple HelloREEF, this is called HellREEFYarn, that contains more Yarn related API calls. It also shows user an example for how to call those APIs.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs [hunk 5]
```diff
@@ -115,6 +121,39 @@ private void LogApplicationReport()
...
+        {
+            try
+            {
+                var state = _reefClient.KillJobApplication(appId);
+                Logger.Log(Level.Info, "Application killed with state: ", state);
+            }
+            catch (Exception e)
+            {
+                Logger.Log(Level.Error, "Fail to kill Application with exception:", e);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: Why is it safe to swallow this exception? Shouldn't the program crash here? Or, if the kill can fail, shouldn't the call return a `boolean` whether or not it succeeded?

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: I'd prefer for us to make that a bit more user friendly. Maybe catch the exceptions and throw them in a `ApplicationKillFailedException` or so that one can interrogate as to why it failed?

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs [hunk 6]
```diff
@@ -33,5 +34,14 @@ public interface IYarnREEFClient : IREEFClient
...
         Task<IReadOnlyDictionary<string, IApplicationReport>> GetApplicationReports();
+
+        /// <summary>
+        /// Kills the job application and returns Job status
+        /// If Application Id is invalid or application has been completed, exception will be thrown.
+        /// </summary>
+        /// <param name="appId">Application id to kill.</param>
+        /// <returns></returns>
+        [Unstable("0.17", "Working in progress for rest API status returned")]
+        Task<FinalState> KillJobApplication(string appId);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: As I said in the separate comment: This needs some error conditions. Specifically, define the `returns` in the xml doc and define what exceptions are thrown by this method.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs [hunk 7]
```diff
@@ -95,10 +96,15 @@ private void Run()
                 .Build();
 
             var result = _reefClient.SubmitAndGetJobStatus(helloJobRequest);
-
             LogApplicationReport();
+
+            //// This is to test Kill Job Application
+            //// KillJobApplication(result.AppId);
+
+            LogApplicationReport(result.AppId);
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: So we call `LogApplicationReport()` twice?

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs [hunk 8]
```diff
@@ -115,6 +121,39 @@ private void LogApplicationReport()
...
+        /// Get application report and log for given appId
+        /// </summary>
+        private void LogApplicationReport(string appId)
+        {
+            Logger.Log(Level.Info, "Getting Application report...");
+            var apps = _reefClient.GetApplicationReports().Result;
+            Logger.Log(Level.Info, "Got Application report: {0}", apps.Count);
+            IApplicationReport report;
+            apps.TryGetValue(appId, out report);
+            if (report != null)
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: replace lines 131-133 with
```C#
if (apps.TryGetValue(appId, out var report))
{
    // ...
```

Comment (motus) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: (Can check for logging level as well)

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs (hunk 7 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: It was to check the app status after killing it. Also to test LogApplicationReport with appId. previous call doesn't have appId. As KillJobApplication() is commented. I can remove this call.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: Right, should directly use if (apps.TryGetValue(appId, out var report))...

What do you mean check logging level? I noticed in Java code, we specify log level in Log(...), we also check if log level is ..., then log. But in C#, we usually don't do it. Because if the current log level is lower than the one in log statement, it won't be logged.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: The catch can be removed. It is up to the client how they want to handle the exception.

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: Done

**Comment by jwang98052** [comment]
@motus Have addressed review comments and pushed the change. Thanks!

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs [hunk 9]
```diff
@@ -36,13 +36,11 @@ public interface IYarnREEFClient : IREEFClient
...
-        /// If application id is invalid, or application has been completed throw InvalidOperationException.
-        /// If rest request is not accepted, throw YarnRestAPIException.
+        /// Kills the job application
         /// </summary>
         /// <param name="appId">Application id to kill.</param>
-        /// <returns>It returns the Job Final State.</returns>
+        /// <returns>Returns true if the application is killed otherwise return false.</returns>
         [Unstable("0.17", "Working in progress for rest API status returned")]
-        Task<FinalState> KillJobApplication(string appId);
+        bool KillJobApplication(string appId);
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: Just a question: can one REEF client be connected to more than one YARN application at the same time? If it cannot, I don't understand why the user should specify an `appId` explicitly in the kill call.

Comment (motus) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: Also, I've just realized how confusing the method name is. Do we kill a **job**, or an **application**, or an application for a job? maybe just call it `.KillJob()`?

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: Theoretically one client can submit multiple applications. So we need to specify application id for the call.

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: It can be called KillApplication though.

**Comment by jwang98052** [comment]
I have changed the API to return bool which indicate if kill is successful or not. I do think user need examples for Yarn related APIs. There are related questions asked by our clients again and again. have example would help a lot. 
Example usually shows happy scenarios. For test invalid cases, yes, we should put them into test cases. However, .Net now doesn't run test on Yarn as we still have some unresolved issues. We have REEF-1541 opened for it. Also Yarn cluster doesn't have test framework set up for running the .Net test. We should add more Yarn tests after the issues are resolved.

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 9 above)
Comment (motus) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: I am fine with `.KillApplication()` (or even just `.Kill()`).

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs [hunk 10]
```diff
@@ -33,5 +34,13 @@ public interface IYarnREEFClient : IREEFClient
...
         [Unstable("0.17", "Working in progress for rest API id returned")]
         Task<IReadOnlyDictionary<string, IApplicationReport>> GetApplicationReports();
+
+        /// <summary>
+        /// Kills the job application
+        /// </summary>
+        /// <param name="appId">Application id to kill.</param>
+        /// <returns>Returns true if the application is killed otherwise return false.</returns>
+        [Unstable("0.17", "Working in progress for rest API status returned")]
+        bool KillJobApplication(string appId);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: As discussed elsewhere, this should probably be renamed to `Kill` or `KillApplication`. Also, the documentation should be updated to reflect that.

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: Would it make sense to make this method `async`?

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/YarnClient.cs [hunk 11]
```diff
@@ -145,6 +146,46 @@ internal sealed partial class YarnClient : IYarnRMClient
...
+        /// </summary>
+        /// <param name="appId">The application identifier.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>Returns true if the application is killed otherwise returns false.</returns>
+        /// <exception cref="YarnRestAPIException"></exception>
+        public async Task<bool> KillApplicationAsync(string appId, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var killApplication = new KillApplication();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/YarnClient.cs]: Maybe use an object initializer here?

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 10 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: yes

File: lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/YarnClient.cs (hunk 11 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/YarnClient.cs]: sure

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 10 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: changed

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFClient.cs]: renamed

**Comment by jwang98052** [comment]
@motus @markusweimer Have addressed your review comments and pushed in the change.
